### PR TITLE
[Hiếu] Chat Detail

### PR DESCRIPTION
**Pull request by HieuJC**

### CODE REVIEW

File: app/src/main/java/com/hieuminh/chatapp/presenter/contract/ChatDetailContract.kt [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import com.hieuminh.chatapp.model.MessageModel
+import com.hieuminh.chatapp.model.WordModel
+import com.hieuminh.chatapp.presenter.BasePresenter
+import com.hieuminh.chatapp.ui.adapter.BaseAdapter
+
+class ChatDetailContract {
+
+    interface Presenter : BasePresenter {
+        fun getMessage(chatId: String)
+        fun send(word : WordModel, chatId : String)
```
Comment (vuphu-jc) [inline app/src/main/java/com/hieuminh/chatapp/presenter/contract/ChatDetailContract.kt]:
```suggestion
        fun sendMessage(word : WordModel, chatId : String)
```

File: app/src/main/java/com/hieuminh/chatapp/presenter/impl/ChatDetailPresenterImpl.kt [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    private val mDataBaseReference = FirebaseDatabase.getInstance().reference
+    private val mAuth: FirebaseAuth = FirebaseAuth.getInstance()
+    private var messageModel: MessageModel? = null
+
+    override fun getMessage(chatId: String) {
+        mDataBaseReference.child(CHAT).child(chatId).addValueEventListener(object :
+            ValueEventListener {
+            override fun onDataChange(snapshot: DataSnapshot) {
+                val message = snapshot.getValue(MessageModel::class.java)
+                print(message?.messages.toString())
```
Comment (vuphu-jc) [inline app/src/main/java/com/hieuminh/chatapp/presenter/impl/ChatDetailPresenterImpl.kt]:
```suggestion
```

File: app/src/main/java/com/hieuminh/chatapp/ui/MainActivity.kt [hunk 3]
```diff
@@ -10,4 +10,7 @@ class MainActivity : AppCompatActivity() {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
     }
+
+    override fun onBackPressed() {
+    }
```
Comment (vuphu-jc) [inline app/src/main/java/com/hieuminh/chatapp/ui/MainActivity.kt]: remove if not using

File: app/src/main/java/com/hieuminh/chatapp/utils/TimeUtils.kt [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package com.hieuminh.chatapp.utils
+
+import java.util.*
+
+fun Date?.toStr() : String {
+    return "${this?.hours}:${this?.minutes}:${this?.seconds} ${this?.day}/${this?.month}/${this?.year}"
```
Comment (chuongdevJC) [inline app/src/main/java/com/hieuminh/chatapp/utils/TimeUtils.kt]: define a const

File: app/src/main/java/com/hieuminh/chatapp/ui/adapter/ChatHomeAdapter.kt [hunk 5]
```diff
@@ -44,16 +46,23 @@ class ChatHomeAdapter() :
...
-            val lastWord: WordModel? = messageModel.messages?.last()
+            val ownerUser: UserModel?
+            val partnerUser : UserModel?
+                if(mAuth.uid == messageModel.firstUser?.id) {
+                    ownerUser = messageModel.firstUser
+                    partnerUser = messageModel.secondUser
+                } else {
+                    ownerUser = messageModel.secondUser
+                    partnerUser = messageModel.firstUser
+                }
```
Comment (vuphu-jc) [inline app/src/main/java/com/hieuminh/chatapp/ui/adapter/ChatHomeAdapter.kt]: format code

File: app/src/main/java/com/hieuminh/chatapp/ui/home/AddChatFragment.kt [hunk 6]
```diff
@@ -77,15 +80,15 @@ class AddChatFragment : Fragment(), AddChatContract.View {
...
     override fun onStartProcessBar() {
-        if (llSearchNotFound.visibility != GONE)
-            llSearchNotFound.visibility = GONE
-        if (llSearchSuccess.visibility != GONE)
-            llSearchSuccess.visibility = GONE
-        processBar.visibility = VISIBLE
+        if (mView.llSearchNotFound.visibility != GONE)
+            mView.llSearchNotFound.visibility = GONE
+        if (mView.llSearchSuccess.visibility != GONE)
+            mView.llSearchSuccess.visibility = GONE
```
Comment (vuphu-jc) [inline app/src/main/java/com/hieuminh/chatapp/ui/home/AddChatFragment.kt]:
```suggestion
        mView.llSearchNotFound.visibility = GONE
        mView.llSearchSuccess.visibility = GONE
```

File: app/src/main/java/com/hieuminh/chatapp/ui/home/ChatDetailFragment.kt [hunk 7]
```diff
@@ -1,22 +1,122 @@
...
-        super.onCreate(savedInstanceState)
-    }
+    private lateinit var mChatDetailNavigator: ChatDetailNavigator
+    private lateinit var mChatDetailAdapter: ChatDetailAdapter
+    private lateinit var mChatDetailPresenter: ChatDetailContract.Presenter
+    private val mAuth = FirebaseAuth.getInstance()
+    private var ownerUser: UserModel? = null
+    private var partnerUser: UserModel? = null
+    private lateinit var chatId: String
+    private lateinit var mView: View
```
Comment (vuphu-jc) [inline app/src/main/java/com/hieuminh/chatapp/ui/home/ChatDetailFragment.kt]:
```suggestion
    private lateinit var mChatDetailNavigator: ChatDetailNavigator
    private lateinit var mChatDetailAdapter: ChatDetailAdapter
    private lateinit var mChatDetailPresenter: ChatDetailContract.Presenter
    private lateinit var chatId: String
    private lateinit var mView: View
    private val mAuth = FirebaseAuth.getInstance()
    private var ownerUser: UserModel? = null
    private var partnerUser: UserModel? = null
```

File: app/src/main/java/com/hieuminh/chatapp/ui/home/ChatDetailFragment.kt [hunk 8]
```diff
@@ -1,22 +1,122 @@
...
+        mView.tvName.text = partnerUser?.name
+        mView.rv_chat.scrollToPosition(0)
+    }
+
+    override fun onGetMessageFailure(errorMessage: String) {
+        CustomAlertDialog(requireContext()).startAlertDialog(errorMessage)
+    }
+
+    override fun onSendSuccess() {
+        mView.etText.setText("")
```
Comment (vuphu-jc) [inline app/src/main/java/com/hieuminh/chatapp/ui/home/ChatDetailFragment.kt]: can you use method `clear` instead?

File: app/src/main/java/com/hieuminh/chatapp/ui/home/ChatDetailFragment.kt [hunk 9]
```diff
@@ -1,22 +1,122 @@
...
+        mView.rv_chat.layoutManager = layoutManager
+        mView.rv_chat.setHasFixedSize(true)
+        mView.rv_chat.adapter = mChatDetailAdapter
+    }
+
+    private fun init(view: View) {
+        mChatDetailNavigator = ChatDetailNavigatorImpl(view)
+        initAdapter()
+        initRecyclerView()
+        mChatDetailPresenter = ChatDetailPresenterImpl(this)
```
Comment (vuphu-jc) [inline app/src/main/java/com/hieuminh/chatapp/ui/home/ChatDetailFragment.kt]:
```suggestion
        mChatDetailNavigator = ChatDetailNavigatorImpl(view)
        mChatDetailPresenter = ChatDetailPresenterImpl(this)
        initAdapter()
        initRecyclerView()
```

**Review by AnhJobChat — APPROVED** [review]

**Review by chuongdevJC — APPROVED** [review]

**Review by vuphu-jc — APPROVED** [review]

**Pull request merged**